ref/ruby/bech32.rb

```diff
@@ -0,0 +1,85 @@
+# Copyright (c) 2017 Shigeyuki Azuchi
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+
+module Bech32
+
+  SEPARATOR = '1'
+
+  CHARSET = %w(q p z r y 9 x 8 g f 2 t v d w 0 s 3 j n 5 4 k h c e 6 m u a 7 l)
+
+  module_function
+
+  # Encode Bech32 string
+  def encode(hrp, data)
+    checksummed = data + create_checksum(hrp, data)
+    hrp + SEPARATOR + checksummed.map{|i|CHARSET[i]}.join
+  end
+
+  # Decode a Bech32 string and determine hrp and data
+  def decode(bech)
+    # check uppercase/lowercase
+    return nil if (bech.downcase != bech && bech.upcase != bech)
+    bech.each_char{|c|return nil if c.ord < 33 || c.ord > 126}
+    bech = bech.downcase
+    # check data length
+    pos = bech.rindex(SEPARATOR)
+    return nil if pos.nil? || pos + 7 > bech.length || bech.length > 90
```

Comment:
It looks like the original Python disallows pos == 0 (a bech string starting with a separator) and this code does not?